rig/routing_table/utils.py

```diff
@@ -0,0 +1,238 @@
+            entry_0 = entry_cls(entry.route, entry.key, entry.mask | bit)
+            for new_entry in expand_entry(entry_0, ignore_xs, entry_cls):
+                yield new_entry
+
+            # And yield all the entries with this bit set as 1
+            entry_1 = entry_cls(entry.route, entry.key | bit, entry.mask | bit)
+            for new_entry in expand_entry(entry_1, ignore_xs, entry_cls):
+                yield new_entry
+
+            # Stop looking for Xs
+            break
+    else:
+        # If there are no Xs then yield the entry we were given.
+        yield entry
+
+
+def expand_entries(entries, entry_cls=RoutingTableEntry, ignore_xs=None):
+    """Turn all Xs which are not common to all entries into ``0``s and ``1``s.
+
+    For example::
+
+        >>> from rig.routing_table import RoutingTableEntry
+        >>> entries = [
+        ...     RoutingTableEntry(None, 0b0100, 0xfffffff0 | 0b1100),  # 01XX
+        ...     RoutingTableEntry(None, 0b0010, 0xfffffff0 | 0b0010),  # XX1X
+        ... ]
+        >>> set(expand_entries(entries)) == {
+        ...     RoutingTableEntry(None, 0b0100, 0xfffffff0 | 0b1110),  # 010X
+        ...     RoutingTableEntry(None, 0b0110, 0xfffffff0 | 0b1110),  # 011X
+        ...     RoutingTableEntry(None, 0b0010, 0xfffffff0 | 0b1110),  # 001X
+        ...     RoutingTableEntry(None, 0b0110, 0xfffffff0 | 0b1110),  # 011X
+        ...     RoutingTableEntry(None, 0b1010, 0xfffffff0 | 0b1110),  # 101X
+        ...     RoutingTableEntry(None, 0b1110, 0xfffffff0 | 0b1110),  # 111X
+        ... }
+        True
+
+    Note that the ``X`` in the LSB was retained because it is common to all
+    entries.
+
+    Any duplicated entries will be removed (in this case the first and second
+    entries will both match ``0000``, so when the second entry is expanded only
+    one entry is retained)::
+
+        >>> entries = [
+        ...     RoutingTableEntry(0, 0b0000, 0b1111),  # 0000 -> 0
+        ...     RoutingTableEntry(1, 0b0000, 0b1011),  # 0X00 -> 1
+        ... ]
+        >>> set(expand_entries(entries)) == {
+        ...     RoutingTableEntry(0, 0b0000, 0b1111),  # 0000 -> 0
+        ...     RoutingTableEntry(1, 0b0100, 0b1111),  # 0100 -> 1
+        ... }
+        True
+
+    .. warning::
+
+        The routing table that is returned is guaranteed to be orthogonal, this
+        means that each key that it would match is matched by only one entry.
+        Should the input table be non-orthogonal a warning will be raised
+        indicating which key(s) could be matched by multiple entries.
+
+    Parameters
+    ----------
+    entries : [:py:class:`~rig.routing_table.RoutingTableEntry`...] or similar
+        The entries to expand.
+
+    Other Parameters
+    ----------------
+    entry_cls :
+        Function to use to make new routing table entries. Must have the same
+        signature as :py:class:`~rig.routing_table.RoutingTableEntry` (the
+        default).
+    ignore_xs : int
```

Comment:
Is this ever used? If not I'd suggest removing it since if a user supplies a mask with any non-common bits ignored the orthogonality check will not be reliable.